lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1252,17 +1261,22 @@ def _process_logging_relation_changed(self, relation: Relation):
         Args:
             relation: the `Relation` instance to update.
         """
+        self._update_relation_data(relation)
+        self._update_alert_rules(relation)
+        self._check_alert_rules()
```

Comment:
It definitely seems that `_check_alert_rules()` should simply be called from `_update_alert_rules()`. If it is not, there is a possibility that the rules are _not_ rechecked after generation during `relation-departed`

Reply:
When a LokiPushAPI consumer departs al that is required is to remove the alert rules corresponding to that consumer. There is no need to recheck all the existing rules which were already checked when they were inserted in the first place.

Reply:
`_check_alert_rules()` verifies "is Loki accessible?", and that is still absolutely necessary.
 
 If, for some reason, removing the alert rules put Loki into an error state, we would end up with a charm in `ActiveStatus` with a bad workload. This should not happen.
 
 Conversely, if the charm _was_ in `BlockedStatus` due to bad alert rules, a successful check will still send an event back to the charm which clears that status and puts it into `ActiveStatus` if the check is ok.
 
 Skipping this step is not optional if we want `juju status` to have any meaning.

Reply:
- Issue #149 was created to address the problem of checking "is Loki Accessible".
- Please do create additional issues for suggested improvements. This PR is not meant to be that last one there ever will be with Loki improvements.

Reply:
I think that PR is an improvement is great. But this PR also potentially breaks the Loki workflow.

Picture the following:

* Deploy Loki
* Deploy another charm with "good" alert rules, and a charm with "bad" alert rules
* Relate the charm with "good" alert rules. `_check_alert_rules()` happens, a ` LokiPushApiAlertRulesChanged` event is emitted, the charm gets it, and the `error` state is null, so it stays active.
* Relate the charm with the "bad" alert charm. `_check_alert_rules()` fails, a ` LokiPushApiAlertRulesChanged` event is emitted, the charm gets it, the `error` state is True, so it puts the charm in `BlockedStatus` with the message.
* Remove the relation to the "bad" rules charm. With this PR, `_check_alert_rules()` is never fired, no event is emitted, Loki is not checked, and the charm stays in `BlockedStatus()` with a message about bad alert rules from a relation that doesn't exist anymore.
* `juju status` and the state of the charm do not reflect the reality.

It's not about whether or not we should need tenacity or the "check the endpoint constantly in `charm.py`" part, this is just basic workflow, and this PR breaks it unless `_check_alert_rules()` is also checked on departed/broken.

Reply:
You made a solid point about `juju status` @rbarry82.!

`_update_alert_rules()`  is only called [here](https://github.com/canonical/loki-k8s-operator/blob/cc07f57faa1a15c0ad62d87ae3edb4df109aecf0/lib/charms/loki_k8s/v0/loki_push_api.py#L1250), and [here](https://github.com/canonical/loki-k8s-operator/blob/cc07f57faa1a15c0ad62d87ae3edb4df109aecf0/lib/charms/loki_k8s/v0/loki_push_api.py#L1265)

What about always calling `_check_alert_rules()` **after** `_update_alert_rules()`? This way we will not break the single responsibility principle.... my two cents

Reply:
This problem can not be solved without addressing issues such as #149 , [PR 135](https://github.com/canonical/loki-k8s-operator/pull/135) and a couple of others. My point is *not* that this PR fixes all issues. It is that we need to put our integration tests in a sane state so we can start addressing all the issues including this one. It has been half a month since we started trying to unblock Loki PRs.

Reply:
We got to this point because we were happy to merge PRs when integration tests were Red but now we insist on the most perfect PR that solves all the problems even if integration tests are passing.

Reply:
What? No.

We are not talking about the same problem.

We cannot merge a patch which fundamentally breaks the workflow of Loki while we wait for other issues and PRs for the integration tests.

The ask here is easy: call `_check_alert_rules()` in departed so a `LokiPushApiAlertRulesChanged` is emitted to the charm so a charm which _was_ in `BlockedStatus` can (correctly) go back to `ActiveStatus` if Loki successfully reloads, and so Loki can restart to reload the alert rules (again, I'm not sure if Loki requires this for provisioning)